In [2]:
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [3]:
stocks = ['ONGC.NS', 'RELIANCE.NS', 'IOC.NS', 'BPCL.NS', 'ADANIPORTS.NS', 'GAIL.NS']
df = yf.download(stocks, start='2025-12-01', end='2026-04-21')['Close']

market = yf.download('^NSEI', start='2025-12-01', end='2026-04-21')['Close']

stock_returns = df.pct_change().dropna()
market_returns = market.pct_change().dropna().squeeze()

betas = {}
for stock in stock_returns.columns:
    beta = stock_returns[stock].cov(market_returns) / market_returns.var()
    betas[stock] = round(beta, 2)

print(betas)


[*********************100%***********************]  6 of 6 completed
[*********************100%***********************]  1 of 1 completed

{'ADANIPORTS.NS': 1.62, 'BPCL.NS': 1.32, 'GAIL.NS': 1.2, 'IOC.NS': 1.06, 'ONGC.NS': 0.17, 'RELIANCE.NS': 0.78}


## 📊 Beta Analysis — Geopolitical Crisis Period (Dec 2025 – Apr 2026)

So I calculated the beta of 6 Indian energy and infrastructure stocks during 
the Iran-USA tensions to see which ones were the most sensitive to market moves. 
Honestly, the results were surprising.

---

### What I found:

| Stock | Beta | Risk Level |
|-------|------|------------|
| ADANIPORTS.NS | 1.62 | 🔴 Very High |
| BPCL.NS | 1.32 | 🔴 High |
| GAIL.NS | 1.20 | 🟠 High |
| IOC.NS | 1.06 | 🟠 Moderate |
| RELIANCE.NS | 0.78 | 🟡 Moderate-Low |
| ONGC.NS | 0.17 | 🟢 Very Low |

---

### My Takeaways:

**ADANIPORTS (1.62)** — This one makes sense in hindsight. 
Adani Ports is all about shipping and logistics, and the Iran-USA 
tensions were literally threatening the Strait of Hormuz — one of 
the most important shipping routes in the world. Every escalation 
headline sent this stock flying. Most volatile of the lot.

**BPCL (1.32)** — BPCL buys crude oil and sells petrol and diesel. 
So when crude spikes, their margins get crushed. The market was 
pricing in that fear aggressively every single day during this period.

**GAIL (1.20)** — GAIL imports LNG and the Hormuz disruption 
directly threatened their supply chain. Made sense that the market 
reacted strongly to every news update.

**IOC (1.06)** — Moves almost in line with the market. Government 
backing and price controls gave it some cushion compared to BPCL 
even though both are in the same business.

**RELIANCE (0.78)** — This was interesting. Reliance is not just 
an oil company — it has Jio, retail, and a bunch of other businesses. 
That diversification acted as a natural hedge and kept its beta low 
even during a pure oil crisis.

**ONGC (0.17)** — Biggest surprise for me. ONGC is a pure oil 
producer but had the lowest beta. Turns out being a government PSU 
works as a shield — government intervention keeps the stock stable 
even when the market is going crazy. Also from our earlier analysis 
we saw ONGC was following crude oil prices more than the Nifty during 
this period — so it was living in its own world basically.

### The Big Realisation 💡

I expected ONGC to be the most volatile since it's directly in the 
oil business. But it was actually the safest stock in this group. 
The shipping and logistics play (Adani Ports) was way more reactive 
to the crisis than the actual oil producer. 

This is what makes geopolitical analysis interesting — the indirect 
effects are often bigger than the direct ones.

### What this means for trading:

If I were building an algo strategy around a geopolitical event like this:
- I would **reduce position size** on high beta stocks like ADANI and BPCL
  because they can move against you very fast
- I would **increase position size** on low beta stocks like ONGC and Reliance
  because they are steadier and more predictable
- This concept is called **Beta Hedging** and it is one of the 
  core strategies used in real algorithmic trading